### DISCUSSION TITLE
Not suitable for torrent downloading?

### DISCUSSION THREAD

**Discussion by PeterEmbleton** [seaweedfs/seaweedfs#3014] [General]
Hi Chris, all,

Just testing out SeaweedFS... AND I LOVE IT!!!

but...

I'm not sure if I'm doing something wrong, whether I'v stumbled on a bug, or whether SeaweedFS is just not suitable for the types of file activity that downloading torrents entails...  If I download a torrent directly to a SeaweedFS mount, I get correct filesizes, but corrupted files - from what I've seen, usually about 10% of a file will be corrupt, but I've seen up to 95% of the Ubuntu .iso be corrupt.  Repeatedly verifying file contents and re-downloading the missing parts of the torrent will eventually get me to a valid file, but that's not ideal...

In trouble-shooting, I've determined that:
- The logfiles don't suggest anything is wrong.
- Copying, or scp-ing files onto the same mount works just fine - no file corruption and file checksums are correct.
- Downloading the torrent to a non-SeaweedFS mount and having the torrent client move the file on completion works fine too - no file corruption and file checksums are correct.

Starting from a fresh (ARM-based) system:
`weed server -dir=/mnt/weed`

> I0505 14:03:11 26933 master.go:255] current: 192.168.2.11:9333 peers:
> I0505 14:03:11 26933 file_util.go:23] Folder /mnt/weed Permission: -rwxr-xr-x
> I0505 14:03:11 26933 file_util.go:23] Folder /mnt/weed Permission: -rwxr-xr-x
> I0505 14:03:11 26933 master.go:255] current: 192.168.2.11:9333 peers:192.168.2.11:9333
> I0505 14:03:11 26933 master_server.go:131] Volume Size Limit is 30000 MB
> I0505 14:03:11 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:11 26933 master.go:147] Start Seaweed Master 30GB 3.01 73961e24d8b311b938dfb692f93e172866813487 at 192.168.2.11:9333
> I0505 14:03:11 26933 raft_server.go:119] Starting RaftServer with 192.168.2.11:9333
> I0505 14:03:11 26933 raft_server.go:168] current cluster leader:
> I0505 14:03:12 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:14 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:16 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:18 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:20 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:21 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:23 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:25 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:27 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:29 26933 volume_grpc_client_to_master.go:41] checkWithMaster 192.168.2.11:9333: get master 192.168.2.11:9333 configuration: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 192.168.2.11:19333: connect: connection refused"
> I0505 14:03:29 26933 master.go:196] Start Seaweed Master 30GB 3.01 73961e24d8b311b938dfb692f93e172866813487 grpc server at 192.168.2.11:19333
> I0505 14:03:30 26933 masterclient.go:87] No existing leader found!
> I0505 14:03:30 26933 raft_server.go:190] Initializing new cluster
> I0505 14:03:30 26933 master_server.go:176] leader change event:  => 192.168.2.11:9333
> I0505 14:03:30 26933 master_server.go:179] [ 192.168.2.11:9333 ] 192.168.2.11:9333 becomes leader.
> I0505 14:03:31 26933 disk_location.go:182] Store started on dir: /mnt/weed with 0 volumes max 8
> I0505 14:03:31 26933 disk_location.go:185] Store started on dir: /mnt/weed with 0 ec shards
> I0505 14:03:31 26933 volume_grpc_client_to_master.go:50] Volume server start with seed master nodes: [192.168.2.11:9333]
> I0505 14:03:31 26933 volume.go:364] Start Seaweed volume server 30GB 3.01 73961e24d8b311b938dfb692f93e172866813487 at 192.168.2.11:8080
> I0505 14:03:31 26933 volume_grpc_client_to_master.go:107] Heartbeat to: 192.168.2.11:9333
> I0505 14:03:31 26933 node.go:222] topo adds child DefaultDataCenter
> I0505 14:03:31 26933 node.go:222] topo:DefaultDataCenter adds child DefaultRack
> I0505 14:03:31 26933 node.go:222] topo:DefaultDataCenter:DefaultRack adds child 192.168.2.11:8080
> I0505 14:03:31 26933 node.go:222] topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080 adds child
> I0505 14:03:31 26933 master_grpc_server.go:72] added volume server 0: 192.168.2.11:8080
> I0505 14:03:34 26933 master_grpc_server.go:281] + client .master@192.168.2.11:9333
> I0505 14:03:34 26933 masterclient.go:180] + .master 192.168.2.11:9333 leader:false
> I0505 14:03:40 26933 master_grpc_server.go:281] + client .filer@192.168.2.11:8888
> I0505 14:03:40 26933 masterclient.go:180] + .filer 192.168.2.11:8888 leader:true
> I0505 14:07:33 26933 store.go:133] In dir /mnt/weed adds volume:1 collection: replicaPlacement:000 ttl:
> I0505 14:07:33 26933 volume_loading.go:136] loading index /mnt/weed/1.idx to memory
> I0505 14:07:33 26933 store.go:137] add volume 1
> I0505 14:07:33 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 1
> I0505 14:07:33 26933 volume_layout.go:391] Volume 1 becomes writable
> I0505 14:07:33 26933 volume_growth.go:235] Created Volume 1 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080
> I0505 14:07:33 26933 store.go:133] In dir /mnt/weed adds volume:2 collection: replicaPlacement:000 ttl:
> I0505 14:07:33 26933 volume_loading.go:136] loading index /mnt/weed/2.idx to memory
> I0505 14:07:34 26933 store.go:137] add volume 2
> I0505 14:07:34 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 2
> I0505 14:07:34 26933 volume_layout.go:391] Volume 2 becomes writable
> I0505 14:07:34 26933 volume_growth.go:235] Created Volume 2 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080
> I0505 14:07:34 26933 store.go:133] In dir /mnt/weed adds volume:3 collection: replicaPlacement:000 ttl:
> I0505 14:07:34 26933 volume_loading.go:136] loading index /mnt/weed/3.idx to memory
> I0505 14:07:34 26933 store.go:137] add volume 3
> I0505 14:07:34 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 3
> I0505 14:07:34 26933 volume_layout.go:391] Volume 3 becomes writable
> I0505 14:07:34 26933 volume_growth.go:235] Created Volume 3 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080
> I0505 14:07:34 26933 store.go:133] In dir /mnt/weed adds volume:4 collection: replicaPlacement:000 ttl:
> I0505 14:07:34 26933 volume_loading.go:136] loading index /mnt/weed/4.idx to memory
> I0505 14:07:34 26933 store.go:137] add volume 4
> I0505 14:07:34 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 4
> I0505 14:07:34 26933 volume_layout.go:391] Volume 4 becomes writable
> I0505 14:07:34 26933 volume_growth.go:235] Created Volume 4 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080
> I0505 14:07:34 26933 store.go:133] In dir /mnt/weed adds volume:5 collection: replicaPlacement:000 ttl:
> I0505 14:07:34 26933 volume_loading.go:136] loading index /mnt/weed/5.idx to memory
> I0505 14:07:34 26933 store.go:137] add volume 5
> I0505 14:07:34 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 5
> I0505 14:07:34 26933 volume_layout.go:391] Volume 5 becomes writable
> I0505 14:07:34 26933 volume_growth.go:235] Created Volume 5 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080
> I0505 14:07:34 26933 store.go:133] In dir /mnt/weed adds volume:6 collection: replicaPlacement:000 ttl:
> I0505 14:07:34 26933 volume_loading.go:136] loading index /mnt/weed/6.idx to memory
> I0505 14:07:34 26933 store.go:137] add volume 6
> I0505 14:07:34 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 6
> I0505 14:07:34 26933 volume_layout.go:391] Volume 6 becomes writable
> I0505 14:07:34 26933 volume_growth.go:235] Created Volume 6 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080
> I0505 14:07:34 26933 store.go:133] In dir /mnt/weed adds volume:7 collection: replicaPlacement:000 ttl:
> I0505 14:07:34 26933 volume_loading.go:136] loading index /mnt/weed/7.idx to memory
> I0505 14:07:34 26933 store.go:137] add volume 7
> I0505 14:07:34 26933 volume_grpc_client_to_master.go:158] volume server 192.168.2.11:8080 adds volume 7
> I0505 14:07:34 26933 volume_layout.go:391] Volume 7 becomes writable
> I0505 14:07:34 26933 volume_growth.go:235] Created Volume 7 on topo:DefaultDataCenter:DefaultRack:192.168.2.11:8080

`weed filer -defaultStoreDir=/mnt/weed_filer -master=localhost:9333`

> I0505 14:03:40 26949 filer_server.go:141] default to create filer store dir in /mnt/weed_filer/filerldb2
> I0505 14:03:40 26949 leveldb2_store.go:42] filer store leveldb2 dir: /mnt/weed_filer/filerldb2
> I0505 14:03:40 26949 file_util.go:23] Folder /mnt/weed_filer/filerldb2 Permission: -rwxr-xr-x
> I0505 14:03:41 26949 filer.go:122] create filer.store.id to 1346058999
> I0505 14:03:41 26949 configuration.go:28] configured filer store to leveldb2
> I0505 14:03:41 26949 filer.go:90] the cluster has 1 filers
> I0505 14:03:41 26949 filer.go:246] Start Seaweed Filer 30GB 3.01 73961e24d8b311b938dfb692f93e172866813487 at 192.168.2.11:8888
> I0505 14:03:41 26949 filer_grpc_server_sub_meta.go:246] + listener filer:192.168.2.11:8888@192.168.2.11:52212
> I0505 14:03:41 26949 filer_grpc_server_sub_meta.go:93]  filer:192.168.2.11:8888@192.168.2.11:52212 local subscribe / from 2022-05-05 14:02:41.940184064 +1000 AEST
> I0505 14:03:41 26949 filer_grpc_server_sub_meta.go:105] read on disk filer:192.168.2.11:8888@192.168.2.11:52212 local subscribe / from 2022-05-05 14:02:41.940184064 +1000 AEST
> I0505 14:03:41 26949 filer_grpc_server_sub_meta.go:121] read in memory filer:192.168.2.11:8888@192.168.2.11:52212 local subscribe / from 2022-05-05 14:02:41.940184064 +1000 AEST
> I0505 14:03:57 26949 filer_grpc_server_sub_meta.go:246] + listener mount@127.0.0.1:53064
> I0505 14:03:57 26949 filer_grpc_server_sub_meta.go:33]  mount@127.0.0.1:53064 starts to subscribe / from 2022-05-05 14:03:57.911253984 +1000 AEST

`sudo ./weed mount -filer=localhost:8888 -dir=/mnt/weed_mount -filer.path=/`

> mount point owner uid=1000 gid=1000 mode=drwxr-xr-x
> I0505 14:03:57 26964 leveldb_store.go:47] filer store dir: /tmp/7afe875e/meta
> I0505 14:03:57 26964 file_util.go:23] Folder /tmp/7afe875e/meta Permission: -rwxr-xr-x
> This is SeaweedFS version 30GB 3.01 73961e24d8b311b938dfb692f93e172866813487 linux arm

With this running, I downloaded a file using Transmission.  On completion:

`system@testbox:/mnt/weed_mount/transmission$ ls -la`

> total 1432338
> drwxr-xr-x 1 debian-transmission debian-transmission          0 May  5 14:07 .
> drwxr-xr-x 1 system              system                       0 May  5 14:16 ..
> -rw-r--r-- 1 debian-transmission debian-transmission 1466714112 May  5 14:15 ubuntu-22.04-live-server-amd64.iso
> system@testbox:/mnt/weed_mount/transmission$ echo "84aeaf7823c8c61baa0ae862d0a06b03409394800000b3235854a6b38eb4856f *ubuntu-22.04-live-server-amd64.iso" | shasum -a 256 --check
> ubuntu-22.04-live-server-amd64.iso: FAILED
> shasum: WARNING: 1 computed checksum did NOT match`

In this instance using Transmission to "Verify Local Data", showed that the file was 18.1% complete (i.e. 81.9% of the file was corrupt).

Grateful for your thoughts on this!  I can't imagine I'm the first to have this problem??

**Comment by chrislusf** [comment]
This is a random write pattern. There are different code path optimized for different patterns.

There may be some locking problem.  Need a good way to reproduce this.

**Reply by PeterEmbleton** [reply]
Any suggestions on where I should start looking for problems, or what code paths might be better optimized?

**Reply by chrislusf** [reply]
https://github.com/chrislusf/seaweedfs/blob/master/weed/mount/page_writer/upload_pipeline.go#L80

The `isSequential` mode or not determines different ways to write memory chunks.
This `isSequential` value may possibly switch in the middle in bit-torrent case. Need to add some unit tests for this.

**Comment by chrislusf** [comment]
how large was the file? Please try https://github.com/chrislusf/seaweedfs/releases/tag/3.06 and see whether it helps.
